File: Server_Root/Server/RoomServer/Messages/GameEventMoveToC.cpp
#include "stdafx.h"
#include "GameEventMoveToC.h"
#include "GameEventMoveToC_Data_generated.h"

GameEventMoveToC::GameEventMoveToC() : m_Builder(1024)
{
}

GameEventMoveToC::~GameEventMoveToC()
{
}

unsigned short GameEventMoveToC::GetID()
{
	return MESSAGE_ID;
}

IMessage* GameEventMoveToC::Clone()
{
	return NULL;
}

const char* GameEventMoveToC::Serialize(int* pLength)
{
	FBSData::Vector3 dest(m_vec3Dest.x, m_vec3Dest.y, m_vec3Dest.z);

	GameEventMoveToC_DataBuilder data_builder(m_Builder);
	data_builder.add_PlayerIndex(m_nPlayerIndex);
	data_builder.add_EventTime(m_lEventTime);
	data_builder.add_Dest(&dest);
	auto data = data_builder.Finish();

	m_Builder.Finish(data);

	*pLength = m_Builder.GetSize();

	return (char*)m_Builder.GetBufferPointer();
}

bool GameEventMoveToC::Deserialize(const char* pChar)
{
	auto data = flatbuffers::GetRoot<GameEventMoveToC_Data>((const void*)pChar);

	m_nPlayerIndex = data->PlayerIndex();
	m_lEventTime = data->EventTime();
	m_vec3Dest.x = data->Dest()->x();
	m_vec3Dest.y = data->Dest()->y();
	m_vec3Dest.z = data->Dest()->z();

	return true;
}

File: Client_Root/Client/Assets/Scripts/Network/Messages/ToClient/GameEventMoveToC.cs
using UnityEngine;
using FlatBuffers;

public class GameEventMoveToC : IMessage
{
    public const ushort MESSAGE_ID = MessageID.GameEventMoveToC_ID;

    public int m_nPlayerIndex;
    public long m_lEventTime;
    public Vector3 m_vec3Dest;

    public ushort GetID()
    {
        return MESSAGE_ID;
    }

    public IMessage Clone()
    {
        return null; 
    }

    public byte[] Serialize()
    {
        FlatBufferBuilder builder = new FlatBufferBuilder(1024);

        var dest = FBSData.Vector3.CreateVector3(builder, m_vec3Dest.x, m_vec3Dest.y, m_vec3Dest.z);

        GameEventMoveToC_Data.StartGameEventMoveToC_Data(builder);
        GameEventMoveToC_Data.AddPlayerIndex(builder, m_nPlayerIndex);
        GameEventMoveToC_Data.AddEventTime(builder, m_lEventTime);
        GameEventMoveToC_Data.AddDest(builder, dest);
        var data = GameEventMoveToC_Data.EndGameEventMoveToC_Data(builder);

        builder.Finish(data.Value);

        return builder.SizedByteArray();
    }

    public bool Deserialize(byte[] bytes)
    {
        var buf = new ByteBuffer(bytes);

        var data = GameEventMoveToC_Data.GetRootAsGameEventMoveToC_Data(buf);

        m_nPlayerIndex = data.PlayerIndex;
        m_lEventTime = data.EventTime;
        m_vec3Dest = new Vector3(data.Dest.X, data.Dest.Y, data.Dest.Z);

        return true;
    }
}
